Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,627 @@
+#include "Spire/Ui/TextAreaBox.hpp"
+#include <QAbstractTextDocumentLayout>
+#include <QHBoxLayout>
+#include <QPainter>
+#include <QTextBlock>
+#include <QTextDocument>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Spire/LocalValueModel.hpp"
+#include "Spire/Ui/ScrollableLayer.hpp"
+#include "Spire/Ui/ScrollBar.hpp"
+#include "Spire/Ui/ScrollBox.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  const auto ELLIPSES_CHAR = 0x2026;
+
+  auto DEFAULT_STYLE() {
+    auto style = StyleSheet();
+    auto font = QFont("Roboto");
+    font.setWeight(QFont::Normal);
+    font.setPixelSize(scale_width(12));
+    style.get(Any()).
+      set(BackgroundColor(QColor::fromRgb(0xFF, 0xFF, 0xFF))).
+      set(border(scale_width(1), QColor::fromRgb(0xC8, 0xC8, 0xC8))).
+      set(LineHeight(1.20)).
+      set(TextAlign(Qt::Alignment(Qt::AlignLeft))).
+      set(text_style(font, QColor::fromRgb(0, 0, 0))).
+      set(TextColor(QColor::fromRgb(0, 0, 0))).
+      set(horizontal_padding(8)).
+      set(vertical_padding(5));
+    style.get(Hover() || Focus()).
+      set(border_color(QColor::fromRgb(0x4B, 0x23, 0xA0)));
+    style.get(ReadOnly()).
+      set(BackgroundColor(QColor::fromRgb(0, 0, 0, 0))).
+      set(border_color(QColor::fromRgb(0, 0, 0, 0))).
+      set(horizontal_padding(0));
+    style.get(Disabled()).
+      set(BackgroundColor(QColor::fromRgb(0xF5, 0xF5, 0xF5))).
+      set(border_color(QColor::fromRgb(0xC8, 0xC8, 0xC8))).
+      set(TextColor(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    style.get(ReadOnly() && Disabled()).
+      set(BackgroundColor(QColor::fromRgb(0, 0, 0, 0))).
+      set(border_color(QColor::fromRgb(0, 0, 0, 0)));
+    style.get(Placeholder()).set(TextColor(QColor::fromRgb(0xA0, 0xA0, 0xA0)));
+    style.get(Disabled() / Placeholder()).
+      set(TextColor(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    return style;
+  }
+}
+
+class TextAreaBox::ContentSizedTextEdit : public QTextEdit {
+  public:
+
```

Comment:
Remove vertical space.

Reply:
Do you mean all whitespace between the method definitions?

Reply:
Just line 57.